packages/cw2/README.md

```diff
@@ -0,0 +1,78 @@
+# CW2 Spec: Contract Info for Migration
+
+Most of the CW* specs are focused on the *public interfaces*
+of the contract. The APIs used for `HandleMsg` or `QueryMsg`.
+However, when we wish to migrate from contract A to contract B,
+contract B needs to be aware somehow of how the *state was encoded*.
+
+Generally we use Singletons and Buckets to store the state, but
+if I upgrade to a `cw20-with-bonding-curve` contract, it will only
+work properly if I am migrating from a `cw20-base` contract. But how
+can the new contract know what format the data was stored.
+
+This is where CW2 comes in. It specifies on special Singleton to
+be stored on disk by all contracts on `init`. When the `migrate`
+function is called, then the new contract can read that data and
+see if this is an expected contract we can migrate from. And also
+contain extra version information if we support multiple migrate
+paths.
+
+### Data structures
+
+**Required**
+
+All CW2-compliant contracts must store the following data:
+
+* key: `\x00\x0dcontract_info` (length prefixed "contract_info" using Singleton pattern)
+* data: Json-serialized `ContractInfo`
+
+```rust
+pub struct ContractVersion {
+    /// contract is the crate name of the implementing contract, eg. `crate:cw20-base`
+    /// we will use other prefixes for other languages, and their standard global namespacing
+    pub contract: String,
+    /// version is any string that this implementation knows. It may be simple counter "1", "2".
+    /// or semantic version on release tags "v0.6.2", or some custom feature flag list.
+    /// the only code that needs to understand the version parsing is code that knows how to
+    /// migrate from the given contract (and is tied to it's implementation somehow)
+    pub version: String,
+}
+
+pub struct ContractInfo {
+    versions: Vec<ContractVersion>,
```

Comment:
Can you explain why this is an array?

Reply:
Sure, I was overengineering it.

I will make it simply ContractVersion. That was my second thought last night and thanks for the feedback.